src/Nethermind/Nethermind.Core/JsonConverters/ByteArrayConverter.cs

```diff
@@ -31,31 +31,120 @@ public class ByteArrayConverter : JsonConverter<byte[]>
+        int length = checked((int)valueSequence.Length);
+        if (length == 0) return null;
+        // Detect and skip 0x prefix even if split across segments
+        SequenceReader<byte> sr = new(valueSequence);
+        bool hadPrefix = false;
+        if (sr.TryPeek(out byte b0))
+        {
+            if (b0 == (byte)'0')
+            {
+                sr.Advance(1);
+                if (sr.TryPeek(out byte b1) && (b1 == (byte)'x'))
+                {
+                    sr.Advance(1);
+                    hadPrefix = true;
+                }
+                else
+                {
+                    // rewind if not really a prefix
+                    sr.Rewind(1);
+                }
+            }
+        }
+
+        // Compute total hex digit count (after prefix)
+        long totalHexChars = length - (hadPrefix ? 2 : 0);
+        if (totalHexChars <= 0) return [];
 
-        return returnVal;
+        int odd = (int)(totalHexChars & 1);
+        int outLenFinal = (int)(totalHexChars >> 1) + odd;
+        if (outLenFinal == 0) return [];
+
+        byte[] result = GC.AllocateUninitializedArray<byte>(outLenFinal);
+        Span<byte> output = result;
+        if (odd == 1)
+        {
+            // If odd, we deal with the extra nibble, so we are left with an even number of nibbles
+            if (!sr.TryRead(out byte lastNibble))
+            {
+                ThrowInvalidOperationException();
+            }
+            lastNibble = (byte)HexConverter.FromLowerChar(lastNibble | 0x20);
+            if (lastNibble > 0x0F)
+            {
+                ThrowFormatException();
+            }
+            result[0] = lastNibble;
+            output = output[1..];
+        }
+
+        while (!sr.End)
+        {
+            ReadOnlySpan<byte> first = sr.UnreadSpan;
+            if (!first.IsEmpty)
+            {
+                // Decode the largest even-length slice of the current contiguous span without copying.
+                int evenLen = first.Length & ~1; // largest even
+                if (evenLen > 0)
+                {
+                    int outBytes = evenLen >> 1;
+                    Bytes.FromUtf8HexString(first.Slice(0, evenLen), output.Slice(0, outBytes));
+                    output = output.Slice(outBytes);
+                    sr.Advance(evenLen);
+                    continue;
+                }
+            }
+
+            // Either current span is empty or has exactly 1 trailing nibble; marshal an even-sized chunk.
+            long remaining = sr.Remaining;
+            if (remaining == 0) break;
+
+            // If remaining is even overall, remaining will be >= 2 here; be defensive just in case.
+            if (remaining == 1)
+            {
+                ThrowInvalidOperationException();
+            }
+
+            ushort twoNibbles = 0;
+            Span<byte> temp = MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref twoNibbles, 1));
+            if (!sr.TryCopyTo(temp))
```

Comment:
I'm confused, the line on 114 is suppose to take even number of character from current span. This mean that at this point, there should be exactly 1 character in current span? But you then take two character? Which is still correct since the whole thing in this loop is even. But isnt the point is to operate on current span only?

Reply:
so this is if the chunk is an odd number of hex chars `0x123`, then I have to reach into next chunk to get `34`, using the SequenceReader.TryCopyTo to bridge the spans